src/ray/common/asio/instrumented_io_context.cc

```diff
@@ -22,19 +22,79 @@
 
 #include "ray/common/asio/asio_chaos.h"
 #include "ray/common/asio/asio_util.h"
+#include "ray/stats/metric.h"
+#include "ray/stats/metric_defs.h"
+
+namespace {
+
+// Post a probe. Records the lag and schedule another probe.
+// Requires: `interval_ms` > 0.
+void LagProbeLoop(instrumented_io_context &io_context, int64_t interval_ms) {
+  auto begin = std::chrono::steady_clock::now();
+  io_context.post(
+      [&io_context, begin, interval_ms]() {
+        auto end = std::chrono::steady_clock::now();
+        auto duration =
+            std::chrono::duration_cast<std::chrono::milliseconds>(end - begin);
+        ray::stats::STATS_io_context_event_loop_lag_ms.Record(
+            duration.count(),
+            {
+                {"Name", GetThreadName()},
+            });
+
+        // Schedule the next probe. If `duration` is larger than `interval_ms`, we
+        // should schedule the next probe immediately. Otherwise, we should wait
+        // for `interval_ms - duration`.
+        auto delay = interval_ms - duration.count();
+        if (delay <= 0) {
+          LagProbeLoop(io_context, interval_ms);
+        } else {
+          execute_after(
+              io_context,
+              [&io_context, interval_ms]() { LagProbeLoop(io_context, interval_ms); },
+              std::chrono::milliseconds(delay));
+        }
+      },
+      "event_loop_lag_probe");
+}
+
+void ScheduleLagProbe(instrumented_io_context &io_context) {
+  if (!RayConfig::instance().enable_metrics_collection()) {
+    return;
+  }
+  auto interval =
+      RayConfig::instance().io_context_event_loop_lag_collection_interval_ms();
+  if (interval <= 0) {
+    return;
+  }
+  RAY_LOG(DEBUG) << "Scheduling lag probe for the io_context on thread "
+                 << GetThreadName() << " every " << interval << "ms";
+  // At this time, the `io_context` may not be running yet, so we need to post the
```

Comment:
So the first lag can be very large if the io_context is not running yet?

Should we start the first probe after start?

Reply:
yes that's exactly what we are doing: we post the initial probe